Break the Manifold interface for performance.

drwells opened the pull request:
This is the branch which I have mentioned in passing a few times: we spend nearly a third of our time allocating and freeing memory inside of the Manifold functions called when creating a grid.

I am not sure if we want to proceed with this and there are still a few things that need to be fixed:
- [x] The minimum boost version needs to be increased to ~~1.58~~ 1.59 (update: 1.58 is not compatible with deal.II)
- [x] The dimension-dependent functions for returning the number of default points in each direction are pretty ugly. 
- [x] The `Manifolds::get_default_points_and_weights` function could also be improved.

In addition, I included a commit that adds a new constructor to `ArrayView` which creates an `ArrayView` from a `std::vector`: this allows all tests but one to pass in their current form.

In reference to #4704.

tjhei commented:
/run-tests (I am trying out if the tester gcc-serial is running on demand now)

bangerth commented:
Please submit the `ArrayView` work with a testcase in a separate PR -- this could be merged independently.

What new functionality of BOOST do you need for the current work?

drwells commented:
I only added the new `ArrayView` constructor so that it is possible to call the modified functions with their old arguments. I think it is worth discussing whether or not we should permit that (or make the break even bigger) before moving it to a separate PR.

Regardless of this patch: are people OK with adding a new conversion constructor like this one? I am a bit on the fence as to whether or not we should automatically convert `std::vector`s into `ArrayView`s.

The only new boost functionality that I require it `boost::container::small_vector` which is new in 1.58.

bangerth commented:
I think we found `ArrayView` useful, so making more uses possible seems like the right direction. I would advocate making conversion operators `explicit` to avoid the automatic conversion.

Updating boost requirements is a different issue, of course...

kronbichler commented on include/deal.II/grid/manifold.h:
```diff
@@ -32,6 +33,38 @@ DEAL_II_NAMESPACE_OPEN
+  constexpr int n_default_points_per_cell<3>()
+  {
+    return GeometryInfo<3>::vertices_per_cell + GeometryInfo<3>::lines_per_cell
+           + GeometryInfo<3>::faces_per_cell;
```
Why not simply returning `3^dim-1`?

drwells replied:
I would prefer to write it out this way so that it is clear what the number signifies.

kronbichler commented on include/deal.II/grid/manifold.h:
```diff
@@ -32,6 +33,38 @@ DEAL_II_NAMESPACE_OPEN
+  constexpr int n_default_points_per_cell<3>()
+  {
+    return GeometryInfo<3>::vertices_per_cell + GeometryInfo<3>::lines_per_cell
+           + GeometryInfo<3>::faces_per_cell;
```
I.e., use `Utilities::fixed_int_power<3,dim>-1`

kronbichler commented on source/grid/manifold.cc:
```diff
@@ -144,22 +127,24 @@ get_new_point (const std::vector<Point<spacedim> > &surrounding_points,
+  const std::size_t n_points = surrounding_points.size();
+  // TODO find a better dimension-dependent estimate for the size of this
+  // vector
+  boost::container::small_vector<double, 20> local_weights(n_points);
```
I'm not sure we know that. It really depends on the degree of interpolation (or whatever comes in here). In the end, the allocation is a one-time cost that must be paid off by enough operations. I would have chosen a much larger value here, maybe 200 or 500.

kronbichler commented on source/grid/manifold.cc:
```diff
@@ -659,16 +654,20 @@ add_new_points (const std::vector<Point<spacedim> > &surrounding_points,
+      // The older version used surrounding_points
+      new_points[row] = project_to_manifold(make_array_view(surrounding_points.begin(),
+                                                            surrounding_points.end()),
+                                            new_point);
```
I would use `surrounding_points_start` but it might be that `project_to_manifold` does the necessary adjustments internally. Do we have tests that trigger on one or the other?

drwells replied:
I believe that the answer is 'yes' and our current implementation is wrong. At the same time, we do not use the surrounding points in `FlatManifold::project_to_manifold` (which is another bug in itself; this function should not be the identity function for `structdim != spacedim`) so this is really a separate problem.

drwells commented on source/grid/manifold.cc:
```diff
@@ -659,16 +654,20 @@ add_new_points (const std::vector<Point<spacedim> > &surrounding_points,
+      // The older version used surrounding_points
+      new_points[row] = project_to_manifold(make_array_view(surrounding_points.begin(),
+                                                            surrounding_points.end()),
+                                            new_point);
```
I am not sure; it has been long enough since I wrote this that I don't even really remember the issue. I will look at this with @luca-heltai at some point during the next few days.

drwells commented:
@kronbichler I have made some of the preliminary changes (and I adopted the little optimization you added in a commit to #2418). I will address things more fully once #2418 is merged.

bangerth commented:
Just following up -- what's the status of this patch right now?

drwells commented:
@bangerth What we have now works but there are still some clean-ups I would like to do before merging (see the TODO list at the top of the PR). These are closely related to @kronbichler's criticisms.

I haven't spent much time of this recently; I'd like to finish #4798 before returning to this.

drwells commented:
The first part of this patch has been moved to #5024.

drwells commented:
I would prefer to merge #4798 before we get to this but I will write more patches for this PR this week.

drwells commented:
I addressed most of the review comments and made a few other cosmetic changes. I believe that this is ready to be reviewed again. Here are the main points:
1. I kept, for now, the implicit conversion constructor from a `std::vector` to an `ArrayView`. I added this purely for compatibility: users can still call the old functions with `std::vector`s and things will work. We can make the break even bigger, if we want, by removing this (I would prefer to do that; I don't like implicit conversion constructors).
2. I changed the interface to `TransfiniteInterpolationManifold::compute_chart_points` to take a mutable `ArrayView` as an argument and return a cell iterator. This is not as pretty as the old version (which returns return values instead of populating a provided array) but this approach works better with `ArrayView`. This change aligns with my philosophy '`small_vector` is a memory management detail and should never be in an API'.
3. I think that the implementation of `get_default_points_and_weights` is improved but I would like to hear feedback on it. Perhaps this should be an internal function? How to do this correctly came up in the last review.

After reviewing this and #4798 I don't think it matters which is merged first; this one is probably easier to understand.

kronbichler commented:
I think the `get_default_points_and_weights` function makes sense. It is a bit longish to read but the names are really descriptive so I like it.

kronbichler commented:
/run-tests

kronbichler commented:
Are there any further comments? If nobody objects, I will merge this PR in a day or two.

drwells commented:
If no one feels strongly about preserving the compatibility shim then I will remove the implicit conversion constructor from `std::vector` to `ArrayView`. In addition to deleting 511f57e26d39e2b2d313c41242b6e6748465aa17 I will also have to go back and fix a few dozen tests (which expect the old interface).

drwells commented:
I removed the implicit conversion constructor from a `std::vector` to an `ArrayView`; I can add it back in if anyone would prefer to have it (I would prefer to not add it).

drwells commented on include/deal.II/grid/manifold.h:
```diff
@@ -373,20 +436,18 @@ class Manifold : public Subscriptor
-                 const std::vector<double>           &weights) const;
+  get_new_point (const ArrayView<const Point<spacedim>> &surrounding_points,
+                 const ArrayView<const double>          &weights) const;
+
```
fixed.

kronbichler commented:
I agree that we should not keep the old implicit conversion. It is probably quite rare that users explicitly use the methods in the manifold - usually they only implement the interface, so I guess this should be fine.

kronbichler commented:
@drwells Can you please add a changelog that announces the (incompatible) change? Then this should be good to go.

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -32,13 +33,75 @@ DEAL_II_NAMESPACE_OPEN
+
+  /**
+   * Specialization for hexes (i.e., structure dimension 3): return the
+   * vertices and midpoints of all faces.
```
...and edges

drwells replied:
fixed

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -32,13 +33,75 @@ DEAL_II_NAMESPACE_OPEN
+  inline
+  constexpr std::size_t n_default_points_per_cell()
+  {
+    return std::numeric_limits<std::size_t>::max();
```
There's a cool construct you can use for this general template (which you really don't want to make available to begin with):
```
template<int dim>
constexpr std::size_t n_default_points_per_cell() = delete;
```
This declares the general template (and consequently allows for specializations) but immediately deletes it again.

It's about as useful this way as if you had put a `static_assert(false);` into the body of the function :-)

drwells replied:
This is a neat trick. I remember reading about `=delete`ing general functions and this is exactly the place to do it :)

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -373,20 +436,17 @@ class Manifold : public Subscriptor
-  get_new_point (const std::vector<Point<spacedim> > &surrounding_points,
-                 const std::vector<double>           &weights) const;
+  get_new_point (const ArrayView<const Point<spacedim>> &surrounding_points,
+                 const ArrayView<const double>          &weights) const;
```
Nice. That's going to make so many things simpler and more efficient.

drwells replied:
Yes. As I mentioned in the changelog, we pay a huge price for allocation that we can avoid with this breaking change :)

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -1130,24 +1182,27 @@ namespace Manifolds
+    const auto points_and_weights = get_default_points_and_weights(iterator, with_laplace);
+    return Quadrature<MeshIteratorType::AccessorType::space_dimension>
+    ({points_and_weights.first.begin(), points_and_weights.first.end()},
+    {points_and_weights.second.begin(), points_and_weights.second.end()});
```
I'm pretty sure this calls the constructor that takes two `std::vector`s as arguments, but that's entirely not obvious. Would you mind making this cast explicit?

drwells replied:
I explicitly wrote out the constructor calls.

bangerth commented on source/fe/mapping_manifold.cc:
```diff
@@ -213,14 +213,18 @@ MappingManifold<dim,spacedim>::
+  return cell->get_manifold().get_new_point(make_array_view(vertices.begin(),
+                                                            vertices.end()),
+                                            make_array_view(weights.begin(),
+                                                            weights.end()));
```
Should we have another version of `make_array_view` that just takes a `std::array` as argument? (Just like we already have one for `std::vector`.)

drwells replied:
I could go either way on this one, but I have a slight preference towards using the same iterator-based template that works with all array classes instead of writing half a dozen wrappers.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3810,11 +3810,17 @@ add_line_support_points (const typename Triangulation<dim,spacedim>::cell_iterat
-                                      support_point_weights_perimeter_to_interior[0], a);
+              std::array<Point<spacedim>, 2> vertices;
+              vertices[0] = cell->vertex(GeometryInfo<dim>::line_to_cell_vertices(line_no, 0));
+              vertices[1] = cell->vertex(GeometryInfo<dim>::line_to_cell_vertices(line_no, 1));
```
Could you make the variable `const` by brace-initializing the pair? (Not sure if that works for `std::array`, though.)

bangerth replied:
Had to try it out -- it works.

drwells replied:
I really loathe the way astyle indents the brace initializers but we may as well do it. Fixed.

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -396,14 +456,14 @@ class Manifold : public Subscriptor
-                  std::vector<Point<spacedim> >       &new_points) const;
+  add_new_points (const ArrayView<const Point<spacedim>> &surrounding_points,
+                  const Table<2,double>                  &weights,
+                  ArrayView<Point<spacedim>>              new_points) const;
```
The name `add_new_points` for this function is now no longer accurate. It used to be that the new points were *added* to a vector, but now they're just written into. Consequently, I think the function should just be called `get_new_points()`.

The function documentation should also say, somewhere, that the number of elements in `new_points` should equal the number of rows in `weights`.

drwells replied:
I think you mean the number of columns in weights :)

I agree with the name change. Its just one more incompatibility on top of the pile of incompatibilities at this point; we just added this function in 8.5 so I think we will be okay (and anyone who is a hardcore `Manifold` user will have to rewrite their codes anyway).

drwells replied:
Clarification: the documentation already states that `The number of columns in @p weights must match the length of @p new_points`.

bangerth replied:
Unless you've already done it, we can also defer the renaming of `add_new_points()` to a later patch.

drwells replied:
I went ahead and changed it anyway. I'll push changes in a few more minutes.

drwells commented:
I have made the requested changes (including the changelog entry).

bangerth commented on doc/news/changes/incompatibilities/20170910DavidWells:
```diff
@@ -0,0 +1,20 @@
+  <li>Manifold::project_to_manifold</li>
+</ol>
+have had their declarations changed in an incompatible manner: these three
+methods now take arguments that are ArrayView​s instead of
```
well, it's actually only two methods :-)

If you write it as "ArrayView objects", then doxygen will link to the class correctly.

bangerth commented on doc/news/changes/incompatibilities/20170910DavidWells:
```diff
@@ -0,0 +1,20 @@
+eliminates this allocation cost.
+
+The method <code>Manifold::add_new_points</code> has been removed in favor of
+<code>Manifold::get_new_points</code>, which also uses ArrayView​s instead of
```
same here

you don't need to use `<code>` for `Manifold::get_new_points -- doxygen will automatically print it with tt font because it knows about it.

tamiko replied:
... when using `Manifold::get_new_points()` (opening and closing parenthesis).

tamiko commented:
@drwells Please merge after updating the changelog.

Looks good!

drwells commented:
Clang is not happy with the internal function: perhaps it does not like the `= delete` syntax.

bangerth commented:
Ah, what a shame.

I think you can generally define the function as
```
  template <int dim>
  uint ()
  {
  return
     GeometryInfo<dim>::vertices_per_cell + 
     GeometryInfo<dim>::lines_per_cell +
     GeometryInfo<dim>::quads_per_cell  +
     GeometryInfo<dim>::hexes_per_cell 
     - 1;   // don't count the cell itself, just the bounding objects
```
Things like `quads_per_cell` should be zero in dim == 1.

drwells commented:
I cannot locally reproduce the clang problem; I suspect that they have since fixed this bug. I'll upload the fixed version in a moment.

drwells commented:
I have made the requested changes.

@bangerth's proposal is a lot nicer than the list of internal functions, so I am glad we went back and changed this.

drwells commented:
Blah, I screwed something up when rebasing on different machines: hold on.

drwells commented:
All right; this should be good to go now.

bangerth commented:
OK to merge once the tester is happy.
